### PR TITLE
hook up the refactored unlock.min.js

### LINKED ISSUE #4385: refactor unlock-js to be more SOLID
This is an epic to track the PRs for refactoring unlock-js

### PR DESCRIPTION
**Pull request by cellog**
# Description

This is the final piece of the puzzle in the new unlock.min.js

Once committed, it fixes several outstanding bugs.

The file is straightforward, it just instantiates the `IframeHandler`, and uses it to instantiate the pieces (`Wallet`, `MainWindowHandler`, `CheckoutUIHandler`) that make up the unlock.min.js script.

The test takes a birds-eye view. The goal is simply to test whether each of the components are created, and so it tests a single expected behavior of that class. The expectation is that the tests for the separate pieces will verify that each piece implements all of its functionality (and they do, currently). It's a sort of proof-of-correctness-by-induction.

A specific example: when the `CheckoutUIHandler` is instantiated, and its `init()` method is called, when we receive a `READY` message from the checkout iframe, the `CheckoutUIHandler` responds by sending the paywall config with a `CONFIG` message. Thus, to test the correct instantiation of the `CheckoutUIHandler`, all we need to do is test to see if the `CONFIG` message is sent in response to `READY`.

With this pattern, it becomes possible to refactor and the test will still pass. More importantly, it completely eliminates the hassle and also the danger of using jest mocks, which will continue to work even if the file being mocked completely changes.

In any case, since this is a new idea, if it doesn't feel right, the test can be rewritten to use mocks.

screenshots showing key purchases working for ETH lock, ERC20 lock with metamask, and a purchase in safari using user accounts, and showing that user accounts don't pop up without being explicit (again in safari, no wallet present).

NOTE: to do the purchase with user accounts, I had to transfer some DEV tokens to the default account from my metamask account, otherwise it is just the dev standup

out
out
out
out
out

# Issues

Fixes #4385 #4250 #4378

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [x] My code follows the style guidelines of this project, including naming conventions
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [x] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: paywall/src/unlock.js/index.ts [hunk 1]
```diff
@@ -1,16 +1,58 @@
...
+declare const process: {
+  env: {
+    UNLOCK_ENV: 'dev' | 'test' | 'staging' | 'prod'
+    DEBUG: any
+    PAYWALL_URL: string
+    USER_IFRAME_URL: string
+  }
+}
+
+const constants: { [key: string]: StartupConstants } = {
```
Comment (julien51) [inline paywall/src/unlock.js/index.ts]: I am a bit worried by this, but I guess we can move it to its own file and only use the ones we need there?

File: paywall/src/unlock.js/startup.ts [hunk 2]
```diff
@@ -1,38 +1,81 @@
...
+import CheckoutUIHandler from './CheckoutUIHandler'
+import StartupConstants from './startupTypes'
 
-export default function startup(window: UnlockWindow) {
-  // Get the config
-  const normalizedConfig = normalizeConfig(window.unlockProtocolConfig)
+/**
+ * convert all of the lock addresses to lower-case so they are normalized across the app
+ */
+export function normalizeConfig(unlockConfig: any) {
```
Comment (julien51) [inline paywall/src/unlock.js/startup.ts]: I think I have seen that code somewhere else? If that's the case do you mind adding a comment `TODO: DRY-up with ...`?

File: paywall/src/unlock.js/index.ts (hunk 1 above)
Comment (cellog) [reply paywall/src/unlock.js/index.ts]: these are the shared constants that are needed, and the idea was to construct them in `index.ts`, similarly to how the data iframe does things. I don't think it matters much where they live, so a separate file would work fine

File: paywall/src/unlock.js/startup.ts [hunk 3]
```diff
@@ -1,38 +1,81 @@
...
+ */
+export default function startup(
+  window: UnlockWindowNoProtocolYet,
+  constants: StartupConstants
+) {
+  // normalize all of the lock addresses
+  const config = normalizeConfig(window.unlockProtocolConfig)
+  // this next line ensures that the minimally valid configuration is passed to Wallet
+  // TODO: provide some kind of developer mode which lazy-loads more extensive validation
+  if (!config) {
```
Comment (julien51) [inline paywall/src/unlock.js/startup.ts]: 👍

File: paywall/src/unlock.js/startup.ts (hunk 2 above)
Comment (cellog) [reply paywall/src/unlock.js/startup.ts]: it's used on line 42. Where you had seen it before was in `setupPostOffices.ts`, which is going to be deleted in the PR following this one as obsolete, so it's a straight move-for-refactor

**Pull request merged**